src/core/server/elasticsearch/client/log_query_and_deprecation.ts

```diff
@@ -0,0 +1,137 @@
+}
+
+/**
+ * Returns stringified debug information from an Elasticsearch request event
+ * useful for logging in case of an unexpected failure.
+ */
+export function getRequestDebugMeta(event: DiagnosticResult): {
+  url: string;
+  body: string;
+  statusCode: number | null;
+  method: string;
+} {
+  const params = event.meta.request.params;
+  // definition is wrong, `params.querystring` can be either a string or an object
+  const querystring = convertQueryString(params.querystring);
+  return {
+    url: `${params.path}${querystring ? `?${querystring}` : ''}`,
+    body: params.body ? `${ensureString(params.body)}` : '',
+    method: params.method,
+    statusCode: event.statusCode!,
+  };
+}
+
+export const instrumentEsQueryAndDeprecationLogger = ({
+  logger,
+  client,
+  type,
+}: {
+  logger: Logger;
+  client: Client;
+  type: string;
+}) => {
+  const queryLogger = logger.get('query', type);
+  const deprecationLogger = logger.get('deprecation');
+  client.diagnostic.on('response', (error, event) => {
+    if (event) {
+      const opaqueId = event.meta.request.options.opaqueId;
+      const meta = opaqueId
+        ? {
+            http: { request: { id: event.meta.request.options.opaqueId } },
+          }
+        : undefined; // do not clutter logs if opaqueId is not present
+      let queryMsg = '';
+      if (error) {
+        if (error instanceof errors.ResponseError) {
+          queryMsg = `${getResponseMessage(event)} ${getErrorMessage(error)}`;
+        } else {
+          queryMsg = getErrorMessage(error);
+        }
+      } else {
+        queryMsg = getResponseMessage(event);
+      }
+
+      queryLogger.debug(queryMsg, meta);
+
+      if (event.warnings && event.warnings.length > 0) {
+        // Plugins can explicitly mark requests as originating from a user by
+        // removing the `'x-elastic-product-origin': 'kibana'` header that's
+        // added by default. User requests will be shown to users in the
+        // upgrade assistant UI as an action item that has to be addressed
+        // before they upgrade.
+        // Kibana requests will be hidden from the upgrade assistant UI and are
+        // only logged to help developers maintain their plugins
+        const requestOrigin =
+          (event.meta.request.params.headers != null &&
+            (event.meta.request.params.headers[
+              'x-elastic-product-origin'
+            ] as unknown as string)) === 'kibana'
+            ? 'kibana'
+            : 'user';
+
+        // Strip the first 5 stack trace lines as these are irrelavent to finding the call site
+        const stackTrace = new Error().stack?.split('\n').slice(5).join('\n');
+
+        const deprecationMsg = `Elasticsearch deprecation: ${event.warnings}\nOrigin:${requestOrigin}\nStack trace:\n${stackTrace}\nQuery:\n${queryMsg}`;
+        if (requestOrigin === 'kibana') {
+          deprecationLogger.warn(deprecationMsg);
```

Comment:
Can an end-user fix the problem? If not, maybe it shouldn't have the `warning` level? End users are worried about `warning` and `errors` for a good reason - they want their deployment to be healthy. If they can't fix a problem, they will contact us. https://github.com/elastic/kibana/issues/35004 is a reminder that we should pick a log level carefully https://github.com/elastic/kibana/issues/35004

Reply:
that's a good point, we want these logs to surface by default on cloud so we can monitor and fix any deprecations we caused... but `info` would be fine.

Reply:
> but info would be fine

Maybe we can use `debug` and configure Cloud to always produce deprecation logs? 
```yml
logging.loggers:
  - name: elasticsearch.deprecation
    level: all
```

Reply:
> Maybe we can use debug and configure Cloud to always produce deprecation logs?

++ Hadn't considered this, but if the main reason for configuring a higher log level was to ensure they are surfaced on Cloud, then I agree this would simplify things greatly.

Reply:
we wouldn't want to surface deprecations caused by users on cloud by default since this could add a lot of volume. So we would need two logger contexts, `elasticsearch.deprecation.kibana` and `elasticsearch.deprecation.user` so that cloud could only enable debug logs for `elasticsearch.deprecation.kibana`.

In terms of priorities, it's critical that we backport this to 7.16 and give teams a heads-up if they need to action anything. So I'll create an follow-up issue that can be done later to lower the level and change the configuration on cloud.

Reply:
https://github.com/elastic/kibana/issues/120733